### PR TITLE
Create explicit count calls for direct counters in P4_14->16 translation

### PR DESCRIPTION
**Pull request by ChrisDodd**
This is a small piece of Pierce's code, just the part that creates explicit 'count' calls for direct counter externs.

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
We don't have a clear spec for direct counters in the P4-14 spec. I exepceted that they always count, i.e., you don't need to invoke the count method explicitly in each action. Also, it is not clear at all whether such counters are also meant to count misses (e.g., default_action execution).

The standard model (under construction) should specify the counter behavior clearly. Depending on the specification of the behavior of the counters, this code may need to be only inserted in the back-end, and not as P4-16 code. In fact, if the counter extern does not have a count method, then this code should not exist in the IR. Conversely, if the count method exists, we have to answer other important questions about it semantics, e.g., what happens if you call it multiple times in an action, or if you increment multiple counters in the same action.

File: backends/bmv2/jsonconverter.cpp [hunk 1]
```diff
@@ -967,6 +967,12 @@ JsonConverter::convertActionBody(const IR::Vector<IR::StatOrDecl>* body,
                         // Do not generate any code for this operation
                         continue;
                     }
+                } else if (em->originalExternType->name == v1model.directCounter.name) {
+                    if (em->method->name == v1model.directCounter.count.name) {
+                        BUG_CHECK(mc->arguments->size() == 0, "Expected 0 argument for %1%", mc);
+                        // Do not generate any code for this operation
```
Comment (mbudiu-vmw) [inline backends/bmv2/jsonconverter.cpp]: I don't understand this. If you generate no code for the count, then what is the point of having it?

File: frontends/p4/fromv1.0/programStructure.cpp [hunk 2]
```diff
@@ -626,7 +626,8 @@ void ProgramStructure::createDeparser() {
 
 const IR::P4Table*
 ProgramStructure::convertTable(const IR::V1Table* table, cstring newName,
-                               IR::IndexedVector<IR::Declaration>* stateful) {
+                               IR::IndexedVector<IR::Declaration>* stateful,
+                               std::map<cstring, cstring> &mapNames) {
```
Comment (mbudiu-vmw) [inline frontends/p4/fromv1.0/programStructure.cpp]: I would give a more suggestive name to the mapNames argument, perhaps "actionRenameMap".

File: frontends/p4/fromv1.0/programStructure.cpp [hunk 3]
```diff
@@ -1338,7 +1341,8 @@ const IR::Statement* ProgramStructure::convertPrimitive(const IR::Primitive* pri
 
 const IR::P4Action*
 ProgramStructure::convertAction(const IR::ActionFunction* action, cstring newName,
-                                const IR::Meter* meterToAccess) {
+                                const IR::Meter* meterToAccess,
+                                cstring counterToAccess) {
```
Comment (mbudiu-vmw) [inline frontends/p4/fromv1.0/programStructure.cpp]: Why is there always a single counter to access?

File: frontends/p4/fromv1.0/programStructure.cpp [hunk 4]
```diff
@@ -1385,7 +1389,21 @@ ProgramStructure::convertAction(const IR::ActionFunction* action, cstring newNam
         body->push_back(stat);
     }
 
-    // Save the original action name in an annotation.
+    if (counterToAccess != nullptr) {
+        ExpressionConverter conv(this);
+        // add a writeback to the meter field
```
Comment (mbudiu-vmw) [inline frontends/p4/fromv1.0/programStructure.cpp]: Is this comment wrong?

File: frontends/p4/fromv1.0/programStructure.cpp [hunk 5]
```diff
@@ -1385,7 +1389,21 @@ ProgramStructure::convertAction(const IR::ActionFunction* action, cstring newNam
...
+        auto decl = get(counterMap, counterToAccess);
+        CHECK_NULL(decl);
+        auto extObj = new IR::PathExpression(decl->name);
+        auto method = new IR::Member(Util::SourceInfo(), extObj, v1model.directCounter.count.Id());
+        auto args = new IR::Vector<IR::Expression>();
+        auto mc = new IR::MethodCallExpression(Util::SourceInfo(), method,
+                                               emptyTypeArguments, args);
+        auto stat = new IR::MethodCallStatement(mc->srcInfo, mc);
+        body->push_back(stat);
```
Comment (mbudiu-vmw) [inline frontends/p4/fromv1.0/programStructure.cpp]: If there is a return in the body this statement will be unreachable.

File: frontends/p4/fromv1.0/programStructure.cpp [hunk 6]
```diff
@@ -1565,8 +1599,15 @@ ProgramStructure::convertControl(const IR::V1Control* control, cstring newName)
                 ::error("Cannot locate table %1%", c.first->table.name);
                 return nullptr;
             }
-            if (std::find(usedTables.begin(), usedTables.end(), tbl) != usedTables.end())
-                directCounters.emplace(c.first->table.name, c.first->name);
+            if (std::find(usedTables.begin(), usedTables.end(), tbl) != usedTables.end()) {
```
Comment (mbudiu-vmw) [inline frontends/p4/fromv1.0/programStructure.cpp]: What if a counter is used in multiple tables?
What if its count method is invoked multiple times?

**Comment by ChrisDodd** [comment]
That is the point of this change -- to make the behavior of the direct counters explicit.  This is mostly for backends OTHER than bmv2-ss, as bmv2-ss has implicit direct counters (so we end up inserting the count calls, then just removing them again when we create the json).

**Comment by mbudiu-vmw** [comment]
The problem with this change is that you are allowing people to insert explicit calls to count, and then you ignore them.

If you plan to change bmv2 this makes sense, but otherwise for the v1model this is pointless.

File: frontends/p4/fromv1.0/programStructure.cpp (hunk 3 above)
Comment (ChrisDodd) [inline frontends/p4/fromv1.0/programStructure.cpp]: I believe P4_14 only supports a single direct meter and a single direct counter on a table (though it is not really specified).  This code only supports one direct meter and one direct counter per table.

Comment (mbudiu-vmw) [inline frontends/p4/fromv1.0/programStructure.cpp]: I could not see the error message checking for this in the diff. Maybe it's somewhere else...

**Comment by mbudiu-vmw** [comment]
If this is for the purpose of other back-ends this transformation should be done in the back-end, and not in the front-end.

**Comment by mbudiu-vmw** [comment]
To be clear, I actually like this approach of explicitly indicating counter operations better, it's just that this does not seem to be implemented correctly on the only target implementing v1model that comes with the public compiler: bmv2. So people will write code explicitly invoking count (or NOT invoking count), but bmv2 will do the wrong thing in these cases. So it would be nice if BMv2 would be changed to follow this model.

**Comment by ChrisDodd** [comment]
The plan is to switch from using simple_switch and v1model.p4 to using bmv2-psa and psa.p4 (portable switch arch, which is the planned p4_16 standard architecture).  There will be a bunch more changes for that; this is just one small piece that seems relatively self-contained.

File: frontends/p4/fromv1.0/programStructure.cpp (hunk 3 above)
Comment (ChrisDodd) [inline frontends/p4/fromv1.0/programStructure.cpp]: This is a problem with the existing code for both direct_counter and direct_meter that should be addressed.

File: frontends/p4/fromv1.0/programStructure.cpp (hunk 5 above)
Comment (ChrisDodd) [inline frontends/p4/fromv1.0/programStructure.cpp]: This was copied from the direct_meter code which is similarly incorrect.

**Comment by mbudiu-vmw** [comment]
By definition P4-14 can only be executed on v1model.p4...
So the changes to the P4-14 front-end do not apply to psa.p4; the changes should not touch v1model.p4.

**Comment by ChrisDodd** [comment]
P4_14 can be executed on any architecture that can support it.  v1model.p4 is just a stop-gap until we can have the 'proper' P4_16 stdarch -- once that is available, we want to get rid of v1model.p4 and make P4_14->16 translation target the p4_16 standard architecture.

**Comment by mbudiu-vmw** [comment]
This may be impossible, or it may restrict severely what can be in the standard architecture.
P4-14 has a lot of built-in assumptions about the architecture. 
But I will let the architecture design committee sort that out.

File: frontends/p4/fromv1.0/programStructure.cpp (hunk 5 above)
Comment (ChrisDodd) [inline frontends/p4/fromv1.0/programStructure.cpp]: P4_14 does not support return in action bodies, so this problem cannot occur.

**Pull request merged**